### PR TITLE
Dev tween abort

### PR DESCRIPTION
**Pull request by Rollo-Liu**
- Fixed an issue that aborting an invalid animation collection will crash.
- Added a guard for aborting a non-playing animation.

### CODE REVIEW

File: Tween/AnimationCollection.cs [hunk 1]
```diff
@@ -37,7 +37,7 @@ public AnimationCollection Add(params AbstractAnimation[] animations)
 		/// </summary>
 		/// <param name="onComplete">An optional callback invoked when the AnimationCollection is complete</param>
 		/// <param name="onAbort">An optional callback invoked if the AnimationCollection is aborted</param>
-		public override void Play(Action onComplete = null, Action onAbort = null)
```
Comment (FenrirWillow) [inline Tween/AnimationCollection.cs]: Why the change to a non-optional callback on complete? If you want to keep that change, the documentation above the method should change as well.

File: Tween/ParalleledAnimation.cs [hunk 2]
```diff
@@ -13,13 +13,17 @@ protected override void PlayQueuedAnimations()
...
+				for (var i = 0; i < Animations.Count; ++i)
 				{
-					if (!animation.IsValid)
+					var animation = Animations[i];
+					animation.Play(HandleEachAnimationCompleted, () =>
 					{
-						Animations.Remove(animation);
-					}
-				}));
+						if (IsPlaying && !animation.IsValid)
```
Comment (FenrirWillow) [inline Tween/ParalleledAnimation.cs]: Can animations not be paused? Or delayed? Wouldn't this line cause those to be discarded without playing?

File: Tween/SequencedAnimation.cs [hunk 3]
```diff
@@ -25,7 +25,7 @@ protected override void PlayQueuedAnimations()
 				{
 					nextAnimation.Play(HandleCurrentAnimationCompleted, () =>
 					{
-						if (!nextAnimation.IsValid)
+						if (IsPlaying && !nextAnimation.IsValid)
```
Comment (FenrirWillow) [inline Tween/SequencedAnimation.cs]: Same here.

File: Tween/ParalleledAnimation.cs (hunk 2 above)
Comment (Rollo-Liu) [reply Tween/ParalleledAnimation.cs]: If `IsPuased == true` then `IsPlaying` must be true. Otherwise, it will throw.

We don't support delayed animation. (delay animation -- `WaitAnimation` is just an animation with a timer in)

File: Tween/AnimationCollection.cs (hunk 1 above)
Comment (Rollo-Liu) [reply Tween/AnimationCollection.cs]: Because the base class has overload functions like `Play() { this.Play(null); }` It makes the default parameters become pointless.

And yes I'll update the doc.

**Comment by danhalldubit** [comment]
I'd like to see more animation unit tests at some point if anyone gets the time like @kkjamie did for the UpdateList. Especially for the more complicated parts of the system working together, like loops and collections.

File: Tween/SequencedAnimation.cs (hunk 3 above)
Comment (danhalldubit) [inline Tween/SequencedAnimation.cs]: I can see the argument for wanting a `ParalleledAnimation` to continue if one part is invalid, but I don't think a `SequencedAnimation` ever should, especially not silently. I'd prefer if both collections Abort if one part fails, I can't think of any situations that would be good design for it to happen intentionally. At the very least if they do continue, I'd like to know about it with a `Debug.LogWarning`.

File: Tween/AnimationCollection.cs [hunk 4]
```diff
@@ -35,9 +35,9 @@ public AnimationCollection Add(params AbstractAnimation[] animations)
 		/// <summary>
 		/// Starts playback of the AnimationCollection starting at the first animation
 		/// </summary>
-		/// <param name="onComplete">An optional callback invoked when the AnimationCollection is complete</param>
+		/// <param name="onComplete">A callback invoked when the AnimationCollection is complete</param>
 		/// <param name="onAbort">An optional callback invoked if the AnimationCollection is aborted</param>
-		public override void Play(Action onComplete = null, Action onAbort = null)
+		public override void Play(Action onComplete, Action onAbort = null)
```
Comment (danhalldubit) [inline Tween/AnimationCollection.cs]: Why does this require an `onComplete` now rather than optional?

**Comment by danhalldubit** [comment]
I'm OK with being able to call stop on an animation that isn't playing, but because it's not expected behaviour it might be nice to at least log it. It won't really cut down on checks elsewhere, as we still have to check `if (animation != null)` after dropping the `&& IsPlaying` check. We could do with the equivalent of `SafeInvoke` for aborting animations.

**Comment by danhalldubit** [comment]
I've also noticed that you can currently call `Play` on animations that are already playing, and it'll reset the duration and overwrite the callbacks. This could potentially cause some states to hang if they're waiting for either an `onComplete` or `onAbort` callback to be invoked and something else calls `Play` again.

There's also the potential to cause `AnimationCollection`s to get stuck this way, if someone plays one of the cached animations from the collection mid way. I know this is bad usage of the library but it's an example in favour of throwing exceptions or at least logging warnings/errors when things go wrong, or users might not realise when or why they've messed up as we saw with the `Abort` stuff recently.

File: Tween/AnimationCollection.cs (hunk 4 above)
Comment (Rollo-Liu) [reply Tween/AnimationCollection.cs]: Because we have overload function in `AbstractAnimation`:
`Play() { Play(null); }`.

File: Tween/SequencedAnimation.cs (hunk 3 above)
Comment (Rollo-Liu) [reply Tween/SequencedAnimation.cs]: They should continue -- or the whole validation check system will no longer make sense. I added the `IsValid` check for reasons. It gives fewer crashes and in general, improves the stability of the program. For a released production, the very last thing you wanted to see is crashes.

I had the compiler flag idea for a couple of years. Which you define a flag for your build i.e. `#PRODUCTION`, then the validation will do every silently and not to crashing your game. But without the flag, it will crash and help you debug your game. This idea is very hard to apply since DUCK contains so many gadgets and I simply can't add a random flag for an open source library.

For now, I'll add a warning message there.

Comment (Rollo-Liu) [reply Tween/SequencedAnimation.cs]: Before I added the validation system -- which is the lazy system that helps the developers maintain their invalid animations. The `Fancy™Tween` animation system has been complained by people quite often that it crashes a lot.

Then I decided, instead of teaching people how to use this system correctly and how to write complicated code to maintain your animations, I made this validation system.

It's like the Garbage Collection system. You can ignore it, for a while, until you hit the limit. Then you need to code properly and always keep this thing in your mind. Same for this `Fancy™Tween`. You can use it in a very rough way and it still works -- but it doesn't mean you have done it in its very best form. In your case, you integrated the animation system deeply into your game and you starting notice issues. I believe by now we are kinda dug out every potential of the `Fancy™Tween` system. And it's time to think about some hardcore usages and meanwhile, not losing its simplicity for entry-level usages. This is very hard to do so.

**Comment by Rollo-Liu** [comment]
I had some discussions a while ago back to when we making this system. We decided not to do anything for the user that controlling individual animations inside the animation collection.

1. We can't control it. You created all the animations and put them into the collection, there is no way that we can restrict the user not to touch them.

2. We can't stop people being idiotic. Do you want to do it in this hard way? No problem! Use at your own risks!

**Comment by Rollo-Liu** [comment]
Regards to the Unit Tests -- I am more than happy to add some.

This PR fixed a rare crash where aborting an animation collection which including an on-going invalid animation.

File: Tween/SequencedAnimation.cs (hunk 3 above)
Comment (danhalldubit) [reply Tween/SequencedAnimation.cs]: I see where you're coming from and I agree when it's a simple self contained part you're making safer, but some of the uses we're covering are really hard to debug if they're covered up. E.g. animation collections of various classes extending different types of animation, looping options, things hanging depending on callbacks, and the fact that they're all just added to the update function of one update list make it really hard to debug and actually find the source of a problem.

Also a lot of the time with the collections I don't think we'll actually prevent an exception, just delay it or even cause more in this case. Quite often someone will use an animation collection and in the `onComplete` call a method on one of the targets, set the gameobject inactive etc. If the collection silently skips an invalid part in the middle but still calls the `onComplete`, this will now throw whereas if we aborted the whole animation collection it wouldn't. Unless we are now expected to null check every object we use in the callback, but that adds a lot more code and defeats part of the purpose of the `IsValid` stuff anyway.

Comment (kkjamie) [reply Tween/SequencedAnimation.cs]: For me it boils down to this question "how many errors do we want to swallow?"

Chuan quite rightly pointed out that you don't want your production build to crash, but also you definitely do want your dev builds to error. I don't want to mess about with compile flags either so the question we now ask is, Would you rather your production build fails silently, or your dev build is very explicit. I think I prefer explicity here.

Dan mentioned, that even if you attempt to continue after a failure, you are likely just prolonging the failure, since your behaviour is incorrect. You can leave an object in the incorrect state. There is a tendency to think "oh it's just an animation, it's not that important". But the way we use animations at current is very coupled to the logic and flow of the system. The incorrect behaviour can actually cause bugs like not fading out a loading screen, leaving something blocking something else.

None of us want our production games to crash, but the chances are if you are very explicit with the developer during the course of development there are less likely to be hidden bugs in the release. If we swallow errors, even providing warnings it's very easy to ignore them, and then you forget until it causes a hidden bug IN production that is impossible to debug.

I see both sides. but I lean towards throwing, As for auto reference, I think if an animation becomes invalid (due to destroying an object). Just throw and tell them to fix it. You will be surprised how quickly people will learn what is correct and incorrect usage of the system.

**Comment by kkjamie** [comment]
I think it's long overdue unit tests, as we have all mentioned there are so many different combinations of API usages to cover, and defining exactly what happens with each needs to be done

**Comment by Rollo-Liu** [comment]
Had a quick look at the FancyTween Animation system -- there are actually quite a lot of error handling and most of them have an error message at least.

To fix this issue, where we want to crash in Dev mode and create error logs in Production, we should definitely set up a separated branch.

For now, I'll just add some error messages that following the current standards.

File: Tween/ParalleledAnimation.cs [hunk 5]
```diff
@@ -9,21 +11,29 @@ public class ParalleledAnimation : AnimationCollection
...
+				var animation = Animations[i];
+				if (animation.IsValid)
 				{
-					if (!animation.IsValid)
+					animation.Play(HandleEachAnimationCompleted, () =>
 					{
-						Animations.Remove(animation);
-					}
-				}));
+						if (IsPlaying && !animation.IsValid)
```
Comment (danhalldubit) [inline Tween/ParalleledAnimation.cs]: Do we want to check this `!animation.IsValid`? I could be mistaken but I think the only time we would get this callback is if it's not valid anyway, or the overall collection `IsPlaying` was false due to being aborted.

Plus if for some reason an animation was aborted while valid, it would never be removed from the list and the overall AnimationCollection couldn't complete (same for SequencedAnimation, because of `if (++NumberOfAnimationsCompleted == Animations.Count)`). I know we're not trying to cover the individual modification of animations while in a collection, but if the above it true there's no point making it hang for nothing.

Comment (Rollo-Liu) [reply Tween/ParalleledAnimation.cs]: This is `onAbort` callback. Handling if the aborted animation inside is invalid or not.

This is specifically designed for animation collection. If you have a collection inside a collection, and that collection went invalid and aborted, this parent collection will continue to work...

Comment (Rollo-Liu) [reply Tween/ParalleledAnimation.cs]: And don't worry -- as I mentioned all the error handlings will get a refactory.

I just left it as it is. It's not breaking anything at least...

Comment (danhalldubit) [reply Tween/ParalleledAnimation.cs]: Hadn't realised the case with collections within collections was different, you're right 👍
